docs/developers_notes/basis_module.md

```diff
@@ -0,0 +1,130 @@
+
+```
+basis_add = basis_a + basis_b # calls basis_a.__add__(basis_b)
+
+basis_add = basis_a * basis_b # calls basis_b.__mul_(basis_b)
+```
+
+## The Class `neurostatslib.basis.Basis`
+
+### The Public Method `evaluate`
+
+The `evaluate` method checks input consistency and evaluates the basis function at some sample points. It accepts one or more numpy arrays as input, which represent the sample points at which the basis will be evaluated, and performs the following steps:
+
+1. Checks that the inputs all have the same sample size `N`, and raises a `ValueError` if this is not the case.
+2. Checks that the number of inputs matches what the basis being evaluated expects (e.g., one input for a 1-D basis, N inputs for an N-D basis, or the sum of N 1-D bases), and raises a `ValueError` if this is not the case.
+3. Checks that the output of the method does not exceed `gb_limit` GB of memory size. This size limit is specified at class initialization.
+4. Calls the `_evaluate` method on the input, which is the subclass-specific implementation of the basis set evaluation.
+5. Returns a numpy array of dimension `N x number of bases`, with each basis element evaluated at the samples.
+
+### The Public Method `evaluate_on_grid`
+
+The `evaluate_on_grid` method evaluates the basis set on a grid of equidistant sample points. The user specifies the input as a series of integers, one for each dimension of the basis function, that indicate the number of sample points in each coordinate of the grid.
+
+This method performs the following steps:
+
+1. Checks that the number of inputs matches what the basis being evaluated expects (e.g., one input for a 1-D basis, N inputs for an N-D basis, or the sum of N 1-D bases), and raises a `ValueError` if this is not the case.
+2. Checks that the output of the method does not exceed `gb_limit` GB of memory size. This size limit is specified at class initialization.
+3. Calls the subclass-specific `_get_samples` method, which returns equidistant samples over the domain of the basis function. The domain may depend on the type of basis.
+4. Calls the subclass-specific `_evaluate` method.
+5. Returns both the sample grid points and the evaluation output at each grid point.
+
+### Abstract Methods
+
+The `neurostatslib.basis.Basis` class has the following abstract methods, which every non-abstract subclass must implement:
+
+1. `_evaluate`: Evaluates a basis over some specified samples.
+2. `_get_samples`: Returns a tuple of equidistant samples (as flat numpy arrays) for each dimension of the basis function. These samples are used to form a grid over which the basis is evaluated.
+3. `_check_n_basis_min`: Checks the minimum number of basis functions required. This requirement can be specific to the type of basis.
```

Comment:
Looks like for many of the subclasses, this just checks that the value is positive. Is it worth implementing this in the superclass and telling developers they can overwrite it themselves?